### DISCUSSION TITLE
Feature: Portable version with local config file

### DISCUSSION THREAD

**Discussion by setvisible** [setvisible/DownZemAll#193] [Ideas]
Portable version with local config file
- Windows: *.ini local file instead of the Windows Registry
- Linux: ?

-> Add the choice (local or user registry) as an option in Preferences
-> Modify the build scripts to enable the option for the Portable builds, but disable it for other builds.

Plus:
- Add option to define another Youtube-DL .cache folder (see youtube-dl's command line: `--cache-dir DIR`)

Regards

**Comment by smaragdus** [comment]
I was about to open an issue for portable mode but I saw that such an issue already exists so I will just comment on it.

Currently [DownZemAll!](https://setvisible.github.io/DownZemAll/) uses Windows registry for saving settings:

<code>HKEY_CURRENT_USER\Software\Sebastien Vavassori\DownZemAll!</code>

I will appreciate if [DownZemAll!](https://github.com/setvisible/DownZemAll) can save program settings (Windows) in configuration file (for example- CONFIG, INI, JSON, XML, any type of configuration file would be fine) in stead of in Windows registry.

Portabile mode can be implemented in different ways, I will provide several examples below:

1. **command-line argument** - if the executable, **DownZemAll.exe**, is started with a special parameter, for example `DownZemAll.exe -portable`, settings are saved in program folder, not in Windows registry;
2. **portable flag** - special file in program folder (for example `portable` without extension or `portable.txt`) triggers portable mode, when **DownZemAll!** starts it checks program folder for such file and if it is there settings are saved in configuration file in program folder, not in Windows registry;
3. **configuration file** - when **DownZemAll!** starts it checks program folder for a configuration file (for example- <code>DownZemAll!.ini</code>) and if it is there the program saves its settings in this configuration file in program folder, not in Windows registry;
4. **special sub-folder** - when **DownZemAll!** starts it checks program folder for a special sub-folder (for example `Data` or `Settings`) and if it is there it saves settings in configuration file in this sub-folder, not in Windows registry;
5. **option in preferences** - new check box- `Portable Mode`, which, if selected, forces the program to save settings in configuration file in program folder, not in Windows registry;


There are other ways of triggering portable mode but I suppose the ones I suggested above are the most common ones. Any of these would be fine for me but I don't like command-line arguments that much because it happens that the user might forget to start the program with a special parameter and then the settings would be saved in Windows registry.


I always prefer to use portable versions of programs if such versions exist or if programs can be run in portable mode because of several advantages portable mode offers:

- **backup** - portable programs can be easily backed up on external drives in case disaster strikes- hardware failure or OS failure;
- **mobility** - portable programs can be easily transferred from machine to machine with all their data and settings;
- **testing** - portable programs facilitate testing- one can test different versions (stable, alpha, beta) with different settings or even use multiple instances of one and the same version with different settigs since settings are isolated;

In my opinion if portable mode is implemented it will make the program more flexible and mobile. Information about how to use the program in portable mode may be included in the ReadMe help file.
